src/main/java/seedu/mtracker/filemanager/InstrumentDecoder.java

```diff
@@ -0,0 +1,74 @@
+package seedu.mtracker.filemanager;
+
+import seedu.mtracker.instrument.Instrument;
+import seedu.mtracker.instrument.InstrumentManager;
+import seedu.mtracker.instrument.subinstrument.Crypto;
+import seedu.mtracker.instrument.subinstrument.Etf;
+import seedu.mtracker.instrument.subinstrument.Forex;
+import seedu.mtracker.instrument.subinstrument.Stock;
+
+import java.util.List;
+
+public class InstrumentDecoder {
+    public static void readFile(InstrumentManager instrumentManager, List<String> fileData) {
+        fileData.stream()
+                .forEach((line) -> {
+                    String[] textSegment = line.split(";", -1);
+                    switch (textSegment[0]) {
+                    case ("Crypto"):
+                        addCryptoToList(textSegment, instrumentManager);
+                        break;
+                    case ("Stock"):
+                        addStockToList(textSegment, instrumentManager);
+                        break;
+                    case ("Forex"):
+                        addForexToList(textSegment, instrumentManager);
+                        break;
+                    default:
+                        addEtfToList(textSegment, instrumentManager);
+                    }
```

Comment:
can abstract in a separate function like addSavedInstrumentToList
also define the magic values, like SEPARATOR = ";", INSTRUMENT_TYPE_INDEX = 0, CRYPTO_TYPE = "Crypto" and so on